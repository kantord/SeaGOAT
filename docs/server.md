<!-- markdownlint-disable MD046 -->
# SeaGOAT-server

The `seagoat-server` is an integral component of the Seagoat command-line tool
designed to analyze your codebase and create vector embeddings for it.

While it serves as a backend for the command-line tool, also allows you to
use it through HTTP to build your own SeaGOAT-based applications.

## Starting the server

To boot up the server for a specific repository, use:

```bash
seagoat-server start <repo_path> [--port=<custom_port>]
```

* `repo_path` - Path to your Git repository.
* `--port` - (Optional) Run the server on a specific port.

If you don't specify a custom port, a random port will be assigned to your
server. Don't worry, SeaGOAT will still be able to automatically find
the server corresponding to a specific repository.

## Developing with SeaGOAT-server

SeaGOAT-server not only serves as a backend for the SeaGOAT command-line tool
but also offers developers the capability to integrate its functions to build
custom applications.

### Retrieving server information

As SeaGOAT servers only run on one repository at a time, there is a command
provided in order to gather information about all running servers, including
how to access them through HTTP.

To get detailed information about all active SeaGOAT servers in JSON format,
you can utilize the `server-info` command:

```bash
seagoat-server server-info
```

You will receive a response similar to this:

```json
{
    "version": "0.5.3",
    "globalCache": "/home/myuser/.cache/seagoat",
    "globalConfigFile": "/home/myuser/.config/seagoat/config.yml",
    "servers": {
        "/path/to/repository/1": {
            "cacheLocation": {
              "chroma": "/home/myuser/.cache/seagoat/bfe8133b9e871ea1c8498a0"
            },
            "isRunning": true,
            "host": "127.0.0.1",
            "port": "8080",
            "address": "http://127.0.0.1:8080"
        },
        "/path/to/repository/2": {
            "cacheLocation": {
              "chroma": "/home/myuser/.cache/seagoat/fbee39c83bd47a75e2f839"
            },
            "isRunning": false,
            "host": "127.0.0.1",
            "port": "8081",
            "address": "http://127.0.0.1:8081"
        }
    }
}
```

In this output, you can also see information about where databases/caches
related to your projects are stored. `globalCache` is the parent folder of
all the cache directories, and withing each server, you can find an attribute
called `cacheLocation` which contains the path to the cache directory for
each different type of cache associated with that project.

If you want to create a configuration file, you can see the path for it
in the `globalConfigFile` attribute. This depends on your operating system.
You can also create a configuration file for your project. See
[the configuration documentation](configuration.md) for more information.

### Making queries using the API

If you want to build an application using SeaGOAT-server, first you need to
figure out the address of the server you want to connect to.

Once you have the address, you can start making queries to it. For instance,
this is how you'd make a query using `curl` to the server running on
`http://localhost:32835`:

```bash
curl 'http://localhost:32835/query/example'
```

You will receive a response similar to this one:

```json
{
  "results": [
    {
      "path": "tests/conftest.py",
      "fullPath": "/home/user/repos/SeaGOAT/tests/conftest.py",
      "score": 0.6,
      "blocks": [
        {
          "score": 0.21,
          "lines": [
            {
              "score": 0.21,
              "line": 100,
              "lineText": "def very_relevant_function():",
              "resultTypes": [
                "result"
              ]
            }
          ],
          "lineTypeCount": {
            "result": 1
          }
        },
        {
          "score": 0.6,
          "lines": [
            {
              "score": 0.6,
              "line": 489,
              "lineText": " contents=(\"hello()\\n\" * (i % 50)),",
              "resultTypes": [
                "result"
              ]
            },
            {
              "score": 0.84,
              "line": 490,
              "lineText": "     return foo * bar",
              "resultTypes": [
                "result"
              ]
            }
          ],
          "lineTypeCount": {
            "result": 1
          }
        }
      ]
    },
    {
      "path": "tests/test_cli.py",
      "fullPath": "/home/user/repos/SeaGOAT/tests/test_cli.py",
      "score": 0.87,
      "blocks": [... etc ... ]
    },
    ... etc ...
  ],
  "version": "0.26.0"
}
```

#### Understanding the response

The response contains the following information:

* `version` - This is the version of SeaGOAT being used.
* `results` - This is an array containing your results.

Each result inside results has the following data:

* `path` - The (relative) path of the file within the repository.
* `fullPath` - The absolute path to the file in your filesystem.
* `score` - A number indicating how relevant a result is, smaller is better.
* `blocks` - An array of relevant code blocks from this file.

Within each block you will find:

* `lines` - An array of line objects containing:
  * `score` - Relevance score for this line. See `score` above.
  * `line` - The line number in the file where the result was found.
  * `lineText` - The actual text content of that line.
  * `resultTypes` - An array indicating all types of result on this line:
    * `"result"` Means that the line is directly relevant to the query.
    * `"context"` Means that the line was added as a context line.
    * `"bridge"` It is a type of context line that is added within a block
    rather than around it. It is used to merge code blocks when they are
    nearly touching.
* `lineTypeCount` - An object containing a count of all line types within
  the code block. See `resultTypes` for more.
* `score` - A score for the code block overall.
