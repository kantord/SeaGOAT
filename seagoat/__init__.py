__version__ = "0.49.4"
