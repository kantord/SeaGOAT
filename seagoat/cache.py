import hashlib
import os
import pickle
from pathlib import Path
from typing import Generic
from typing import TypeVar

import appdirs

T = TypeVar("T")

CACHE_FORMAT_VERSION = 15


class Cache(Generic[T]):
    def __init__(self, cache_name: str, path: Path, initial_value: T):
        self._path = path
        self.data = initial_value
        self._cache_name = cache_name

    def load(self):
        try:
            with open(self._get_cache_file(), "rb") as cache_file:
                self.data = pickle.load(cache_file)
        except (FileNotFoundError, pickle.UnpicklingError, EOFError):
            pass

    def persist(self):
        with open(self._get_cache_file(), "wb") as cache_file:
            pickle.dump(self.data, cache_file)

    def _get_cache_file(self):
        return self.get_cache_folder() / self._cache_name

    def get_cache_folder(self):
        cache_folder = self._get_cache_root() / self._get_project_hash()
        cache_folder.mkdir(parents=True, exist_ok=True)

        return cache_folder

    def _get_cache_root(self):
        if "RUNNER_TEMP" in os.environ:
            return Path(os.environ["RUNNER_TEMP"])

        return Path(
            appdirs.user_cache_dir(
                "seagoat-pytest" if "PYTEST_CURRENT_TEST" in os.environ else "seagoat"
            )
        )

    def _get_project_hash(self):
        normalized_path = Path(self._path).expanduser().resolve()
        text = f"""
        Cache version: {CACHE_FORMAT_VERSION}
        Normalized path: {normalized_path}
        """

        return hashlib.sha256(text.encode()).hexdigest()
