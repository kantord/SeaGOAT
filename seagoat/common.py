IGNORED_BRANCHES = {"gh-pages"}
SUPPORTED_FILE_TYPES = {
    ".txt",
    ".md",
    ".py",
    ".c",
    ".cpp",
    ".h",
    ".hpp",
    ".ts",
    ".js",
    ".tsx",
    ".jsx",
    ".html",
}
