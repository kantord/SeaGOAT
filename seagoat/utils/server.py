import os
import socket
from pathlib import Path
from typing import Dict
from typing import TypedDict
from typing import Union

import appdirs
import psutil

from seagoat.utils.json_file import get_json_file_contents
from seagoat.utils.json_file import write_to_json_file


class ServerInfo(TypedDict):
    address: str
    host: str
    port: int
    pid: int


class ServerDoesNotExist(Exception):
    pass


def _get_server_data_file_path() -> Path:
    user_cache_dir = Path(appdirs.user_cache_dir("seagoat-servers"))
    user_cache_dir.mkdir(parents=True, exist_ok=True)
    return user_cache_dir / "serverData.json"


def normalize_repo_path(repo_path: Union[str, Path]) -> str:
    return str(os.path.normpath(Path(repo_path).expanduser().resolve()))


def get_servers_info() -> Dict[str, ServerInfo]:
    path = _get_server_data_file_path()
    if not os.path.exists(path):
        write_to_json_file(path, {})

    contents = get_json_file_contents(path)
    if contents is None:
        contents = {}

    for key in list(contents.keys()):
        if not os.path.exists(key):
            del contents[key]

    return contents


def update_server_info(
    repo_path: Union[str, Path], new_server_data: ServerInfo
) -> None:
    servers_info = get_servers_info()
    repo_id = normalize_repo_path(repo_path)
    servers_info[repo_id] = new_server_data

    write_to_json_file(_get_server_data_file_path(), servers_info)


def stop_server(repo_path: Union[str, Path]) -> None:
    servers_info = get_servers_info()
    repo_id = normalize_repo_path(repo_path)
    if repo_id in servers_info:
        server_info = servers_info[repo_id]
        process = psutil.Process(server_info["pid"])

        servers_info.pop(repo_id)
        write_to_json_file(_get_server_data_file_path(), servers_info)

        process.terminate()
        process.wait()
    else:
        raise ServerDoesNotExist(f"Server for {repo_path} does not exist.")


def get_server_info(repo_path: Union[str, Path]) -> ServerInfo:
    servers_info = get_servers_info()
    repo_id = normalize_repo_path(repo_path)

    if repo_id not in servers_info:
        raise ServerDoesNotExist(f"Server for {repo_path} does not exist.")

    server_info = servers_info[repo_id]
    return server_info


def is_server_running(repo_path: Union[str, Path]) -> bool:
    try:
        server_info = get_server_info(repo_path)
        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as socket_obj:
            return (
                socket_obj.connect_ex((server_info["host"], server_info["port"])) == 0
            )
    except ServerDoesNotExist:
        return False


def get_free_port() -> int:
    socket_obj = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    socket_obj.bind(("", 0))
    _, port = socket_obj.getsockname()
    socket_obj.close()

    return port
