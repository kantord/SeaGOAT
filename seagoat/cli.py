import os
from functools import cache

import click
import requests
from pygments import highlight
from pygments.formatters import TerminalFormatter
from pygments.lexers import get_lexer_for_filename
from pygments.lexers.javascript import JavascriptLexer
from pygments.lexers.javascript import TypeScriptLexer

from seagoat.server import get_server_info_file
from seagoat.server import load_server_info


def query_server(query, server_address):
    response = requests.get(f"{server_address}/query/{query}")
    try:
        response.raise_for_status()
    except requests.exceptions.HTTPError as err:
        print("HTTP error occurred:")
        print(f"Response code: {response.status_code}")
        print(f"Error: {err}")
        print(f"Response body: {response.text}")
        raise
    return response.json()["results"]


@cache
def get_highlighted_lines(file_name: str):
    with open(file_name, "r", encoding="utf-8") as source_code_file:
        code = source_code_file.read()

    if file_name.endswith(".md"):
        return code.splitlines()

    if file_name.endswith(".jsx"):
        lexer = JavascriptLexer()
    elif file_name.endswith(".tsx"):
        lexer = TypeScriptLexer()
    else:
        lexer = get_lexer_for_filename(file_name)

    result = highlight(code, lexer, TerminalFormatter())

    return result.splitlines()


def print_result_line(result, line, color_enabled):
    if color_enabled:
        highlighted_lines = get_highlighted_lines(str(result["full_path"]))
        print(
            f"{result['path']}:{click.style(str(line), bold=True)}:{highlighted_lines[line - 1]}"
        )
    else:
        for line_content in result["lines"]:
            if line_content["line"] == line:
                print(f"{result['path']}:{line}:{line_content['line_text']}")
                break


@click.command()
@click.argument("query")
@click.argument("repo_path", required=False, default=os.getcwd())
@click.option(
    "--no-color",
    is_flag=True,
    help="Disable formatting. Automatically enabled when part of a bash pipeline.",
)
def seagoat(query, repo_path, no_color):
    """
    Query your codebase for your QUERY in the Git repository REPO_PATH.

    Your query can either be text that you expect to find in a file,
    or a description of what you are looking for.

    When REPO_PATH is not specified, the current working directory is
    assumed to be the repository path.
    """
    _, __, ___, server_address = load_server_info(get_server_info_file(repo_path))
    results = query_server(query, server_address)

    color_enabled = os.isatty(0) and not no_color
    for result in results:
        for result_line in result.get("lines", []):
            print_result_line(result, result_line["line"], color_enabled)


if __name__ == "__main__":
    # pylint: disable-next=no-value-for-parameter
    seagoat()
