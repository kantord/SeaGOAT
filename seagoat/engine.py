"""
    This module allows you to use seagoat as a library
"""
import asyncio
from concurrent.futures import ThreadPoolExecutor
from functools import partial
from itertools import chain
from pathlib import Path
from typing import Dict
from typing import List
from typing import Set

import nest_asyncio
from tqdm import tqdm
from typing_extensions import TypedDict

from seagoat.cache import Cache
from seagoat.file import File
from seagoat.repository import Repository
from seagoat.sources import chroma
from seagoat.sources import ripgrep


RepositoryData = TypedDict(
    "RepositoryData",
    {
        "last_analyzed_version_of_branch": Dict[str, str],
        "required_commits": Set[str],
        "commits_already_analyzed": Set[str],
        "file_data": Dict[str, File],
        "sorted_files": List[str],
        "chunks_already_analyzed": Set[str],
    },
)


nest_asyncio.apply()


class Engine:
    """
    A search engine for a code repository
    """

    def __init__(self, path: str):
        """
        Initializes the library
        """
        self.path = path
        self.query_string = ""
        self._results_from_chromadb = []
        self._results = []
        self._cache = Cache[RepositoryData](
            "cache",
            Path(path),
            {
                "last_analyzed_version_of_branch": {},
                "required_commits": set(),
                "commits_already_analyzed": set(),
                "file_data": {},
                "sorted_files": [],
                "chunks_already_analyzed": set(),
            },
        )
        self._cache.load()
        self.repository = Repository(path, self._cache)
        self._fetchers = {
            "async": [
                ripgrep.initialize(self.repository),
            ],
            "sync": [
                chroma.initialize(self.repository),
            ],
        }

    def analyze_codebase(self):
        self.repository.analyze_files()
        self._create_vector_embeddings()

    def _add_to_collection(self, chunk):
        for source in chain(*self._fetchers.values()):
            source["cache_chunk"](chunk)

    def _create_vector_embeddings(self):
        chunks_to_process = []
        minimum_files_to_analyze = max(40, int(len(self.repository.top_files()) * 0.2))
        for file in self.repository.top_files()[:minimum_files_to_analyze]:
            for chunk in file.get_chunks():
                chunks_to_process.append(chunk)
        for chunk in tqdm(chunks_to_process, desc="Analyzing source code"):
            if chunk.chunk_id in self._cache.data["chunks_already_analyzed"]:
                continue

            self._add_to_collection(chunk)
            self._cache.data["chunks_already_analyzed"].add(chunk.chunk_id)

        self._cache.persist()

        for source in chain(*self._fetchers.values()):
            source["persist"]()

    def query(self, query: str):
        self.query_string = query

    async def fetch(self):
        self._results = []
        executor = ThreadPoolExecutor(max_workers=1)
        loop = asyncio.get_event_loop()
        async_tasks = [
            loop.run_in_executor(executor, partial(source["fetch"], self.query_string))
            for source in self._fetchers["async"]
        ]

        for source in self._fetchers["sync"]:
            self._results.extend(source["fetch"](self.query_string))

        results = await asyncio.gather(*async_tasks)
        for result in results:
            self._results.extend(result)

    def fetch_sync(self):
        loop = asyncio.get_event_loop()
        loop.run_until_complete(self.fetch())

    def get_results(self):
        merged_results = {}

        for result_item in self._results:
            if result_item.path not in merged_results:
                merged_results[result_item.path] = result_item
                continue

            merged_results[result_item.path].extend(result_item)

        sorted_by_score = sorted(
            merged_results.values(),
            key=lambda x: x.get_best_score(self.query_string),
        )
        position_by_result = {item: i for i, item in enumerate(sorted_by_score)}
        position_by_file = {
            file.path: i for i, file in enumerate(self.repository.top_files())
        }

        return list(
            sorted(
                merged_results.values(),
                key=lambda x: (
                    0.8 * position_by_result[x] + 0.2 * position_by_file[x.path]
                ),
            )
        )
