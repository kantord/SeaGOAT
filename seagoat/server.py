import json
import os
import socket
import time
from importlib.metadata import version
from multiprocessing import Process

import appdirs
import click
from flask import current_app
from flask import Flask
from flask import jsonify
from werkzeug.serving import run_simple

from seagoat.engine import Engine


def create_app(repo_path):
    app = Flask(__name__)
    app.config["PROPAGATE_EXCEPTIONS"] = True

    seagoat_engine = Engine(repo_path)
    seagoat_engine.analyze_codebase()
    app.extensions["seagoat_engine"] = seagoat_engine

    @app.route("/query/<query>")
    def query_codebase(query):
        if "seagoat_engine" not in current_app.extensions:
            raise RuntimeError("seagoat_engine is not initialized")
        current_app.extensions["seagoat_engine"].query(query)
        current_app.extensions["seagoat_engine"].fetch_sync()
        results = current_app.extensions["seagoat_engine"].get_results()
        return {
            "results": [result.to_json() for result in results],
            "version": version("wheel"),
        }

    @app.errorhandler(Exception)
    def handle_exception(exception_to_handle):
        res = {
            "code": 500,
            "error": {
                "type": "Internal Server Error",
                "message": exception_to_handle.message
                if hasattr(exception_to_handle, "message")
                else f"{exception_to_handle}",
            },
        }

        return jsonify(res), 500

    return app


def get_server_info_file(repo_path):
    repo_id = os.path.normpath(repo_path).replace(os.sep, "_")
    user_cache_dir = appdirs.user_cache_dir("seagoat-servers")
    os.makedirs(user_cache_dir, exist_ok=True)
    return os.path.join(user_cache_dir, f"{repo_id}.json")


def start_server(repo_path):
    os.environ["TOKENIZERS_PARALLELISM"] = "false"

    socket_obj = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    socket_obj.bind(("", 0))
    _, port = socket_obj.getsockname()
    socket_obj.close()

    app = create_app(repo_path)
    process = Process(
        target=run_simple, args=("localhost", port, app), kwargs={"use_reloader": False}
    )
    process.start()

    server_info_file = get_server_info_file(repo_path)
    with open(server_info_file, "w", encoding="utf-8") as file:
        json.dump({"host": "localhost", "port": port, "pid": process.pid}, file)


def is_server_running(host, port):
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as socket_obj:
        return socket_obj.connect_ex((host, port)) == 0


def load_server_info(server_info_file):
    with open(server_info_file, "r", encoding="utf-8") as file:
        server_info = json.load(file)
    host = server_info["host"]
    port = server_info["port"]
    pid = server_info.get("pid")
    server_address = f"http://{host}:{port}"
    return host, port, pid, server_address


def wait_for(condition_function, timeout, period=0.05):
    start_time = time.time()
    while not condition_function():
        if time.time() - start_time > timeout:
            raise TimeoutError("Timeout expired while waiting for condition.")
        time.sleep(period)


def get_server(repo_path):
    server_info_file = get_server_info_file(repo_path)

    if os.path.exists(server_info_file):
        host, port, _, server_address = load_server_info(server_info_file)

        if is_server_running(host, port):
            click.echo(f"Server is already running at {server_address}")
            return server_address
        os.remove(server_info_file)

    os.environ["TOKENIZERS_PARALLELISM"] = "true"
    temp_app = create_app(repo_path)
    del temp_app

    start_server(str(repo_path))

    wait_for(lambda: os.path.exists(server_info_file), timeout=60)

    host, port, _, server_address = load_server_info(server_info_file)
    click.echo(f"Server started at {server_address}")
    return server_address


@click.group()
def server():
    pass


@server.command()
@click.argument("repo_path")
def start(repo_path):
    """Starts the server."""
    get_server(repo_path)
    click.echo("Server running.")


def get_status_data(repo_path):
    """Return the status data of the server."""
    server_info_file = get_server_info_file(repo_path)
    status_info = {"isRunning": False, "url": None}

    if os.path.exists(server_info_file):
        with open(server_info_file, "r", encoding="utf-8") as file:
            server_info = json.load(file)

        host = server_info["host"]
        port = server_info["port"]
        pid = server_info.get("pid")
        server_address = f"http://{host}:{port}"

        if is_server_running(host, port):
            status_info = {"isRunning": True, "url": server_address, "pid": pid}

    return status_info


@server.command()
@click.argument("repo_path")
@click.option("--json", "use_json_format", is_flag=True, help="Output status as JSON.")
def status(repo_path, use_json_format):
    """Checks the status of the server."""
    status_info = get_status_data(repo_path)

    if use_json_format:
        click.echo(json.dumps(status_info))
    else:
        if status_info["isRunning"]:
            click.echo(f"Server is running at {status_info['url']}")
        else:
            click.echo("Server is not running.")


@server.command()
@click.argument("repo_path")
def stop(repo_path):
    """Stops the server."""
    server_info_file = get_server_info_file(repo_path)

    if os.path.exists(server_info_file):
        os.remove(server_info_file)

    click.echo(
        "Server stopped. If it was running, it will stop after finishing current tasks."
    )


if __name__ == "__main__":
    server()
