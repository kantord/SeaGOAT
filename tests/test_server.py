# pylint: disable=no-member
import copy
import json
import re
import subprocess
from unittest.mock import ANY

import psutil
import pytest
import requests

from seagoat import __version__
from seagoat.server import get_status_data
from seagoat.server import server as seagoat_server


def normalize_full_paths(data, repo):
    real_repo_path = repo.working_dir
    fake_repo_path = "/path/to/repo"
    deep_copy_of_data = copy.deepcopy(data)

    for result in deep_copy_of_data["results"]:
        result["fullPath"] = result["fullPath"].replace(real_repo_path, fake_repo_path)

    return deep_copy_of_data


def test_query_codebase(server, snapshot, repo):
    query_text = "Markdown"
    url = f"{server}/query/{query_text}"
    response = requests.get(url)

    assert response.status_code == 200, response.text

    data = response.json()
    normalized_data = normalize_full_paths(data, repo)
    normalized_data["version"] = "0.0.0-test"

    assert normalized_data == snapshot
    assert len(data["results"]) > 0
    assert data["version"] == __version__


def test_status_1(repo):
    result = subprocess.run(
        ["python", "-m", "seagoat.server", "status", repo.working_dir],
        capture_output=True,
        text=True,
        check=False,
    )
    assert result.returncode == 0
    assert "Server is not running" in result.stdout


@pytest.mark.usefixtures("server")
def test_status_2(repo):
    result = subprocess.run(
        ["python", "-m", "seagoat.server", "status", repo.working_dir],
        capture_output=True,
        text=True,
        check=False,
    )
    assert result.returncode == 0
    assert "Server is running" in result.stdout


@pytest.mark.usefixtures("server")
def test_stop(repo):
    subprocess.run(
        ["python", "-m", "seagoat.server", "stop", repo.working_dir],
        capture_output=True,
        text=True,
        check=False,
    )
    result = subprocess.run(
        ["python", "-m", "seagoat.server", "status", repo.working_dir],
        capture_output=True,
        text=True,
        check=False,
    )
    assert result.returncode == 0
    assert "Server is not running" in result.stdout


def test_status_with_json_when_server_not_running(repo):
    result = subprocess.run(
        ["python", "-m", "seagoat.server", "status", "--json", repo.working_dir],
        capture_output=True,
        text=True,
        check=False,
    )
    assert result.returncode == 0

    json_result = json.loads(result.stdout)

    assert json_result.get("isRunning") is False
    assert json_result.get("url") is None


@pytest.mark.usefixtures("server")
def test_status_with_json_when_server_running(repo):
    result = subprocess.run(
        ["python", "-m", "seagoat.server", "status", "--json", repo.working_dir],
        capture_output=True,
        text=True,
        check=False,
    )
    assert result.returncode == 0

    json_result = json.loads(result.stdout)

    assert json_result.get("isRunning") is True
    url = json_result.get("url")
    assert re.match(r"http://localhost:\d+", url) is not None


def assert_server_status(repo, running):
    result = subprocess.run(
        ["python", "-m", "seagoat.server", "status", "--json", repo.working_dir],
        capture_output=True,
        text=True,
        check=True,
    )
    json_result = json.loads(result.stdout)
    assert json_result.get("isRunning") == running


def simulate_server_dying(repo):
    pid = get_status_data(repo.working_dir)["pid"]

    process = psutil.Process(pid)
    process.terminate()
    process.wait(timeout=10)


@pytest.mark.usefixtures("server")
def test_server_status_not_running_if_process_does_not_exist(repo):
    assert_server_status(repo, running=True)
    simulate_server_dying(repo)
    assert_server_status(repo, running=False)


@pytest.mark.parametrize("limit_value", [1, 3, 7])
def test_query_with_limit_clue_param(client, limit_value, mock_queue):
    response = client.get(f"/query/Markdown?limitClue={limit_value}")
    mock_queue.enqueue.assert_called_with(
        "query",
        query="Markdown",
        limit_clue=limit_value,
        context_below=ANY,
        context_above=ANY,
    )
    assert response.status_code == 200


@pytest.mark.parametrize("context_above", [0, 7])
def test_query_with_context_above(client, context_above, mock_queue):
    response = client.get(f"/query/Markdown?contextAbove={context_above}")
    mock_queue.enqueue.assert_called_with(
        "query",
        query="Markdown",
        context_above=context_above,
        limit_clue=ANY,
        context_below=ANY,
    )
    assert response.status_code == 200


@pytest.mark.parametrize("context_below", [0, 2])
def test_query_with_context_below(client, mock_queue, context_below):
    response = client.get(f"/query/Markdown?contextBelow={context_below}")
    mock_queue.enqueue.assert_called_with(
        "query",
        query="Markdown",
        context_below=context_below,
        limit_clue=ANY,
        context_above=ANY,
    )
    assert response.status_code == 200


def test_version_option(runner):
    result = runner.invoke(seagoat_server, ["--version"])

    assert result.exit_code == 0
    assert result.output.strip() == f"seagoat, version {__version__}"
