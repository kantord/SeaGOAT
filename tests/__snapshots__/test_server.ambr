# serializer version: 1
# name: test_query_codebase
  dict({
    'results': list([
      dict({
        'blocks': list([
          dict({
            'lineTypeCount': dict({
              'context': 3,
              'result': 3,
            }),
            'lines': list([
              dict({
                'line': 0,
                'lineText': "HTML is for markup, but it's complicated. Markdown is simpler.",
                'resultTypes': list([
                  'context',
                ]),
                'score': 0.0,
              }),
              dict({
                'line': 1,
                'lineText': '# Markdown file',
                'resultTypes': list([
                  'context',
                  'result',
                ]),
                'score': 0.6,
              }),
              dict({
                'line': 2,
                'lineText': 'This is an example Markdown file. Updated. ',
                'resultTypes': list([
                  'context',
                  'result',
                ]),
                'score': 0.6,
              }),
              dict({
                'line': 3,
                'lineText': "HTML is for markup, but it's complicated. Markdown is simpler.",
                'resultTypes': list([
                  'result',
                ]),
                'score': 0.6,
              }),
            ]),
            'score': 0.6000000000000001,
          }),
        ]),
        'fullPath': '/path/to/repo/file1.md',
        'path': 'file1.md',
        'score': 0.9,
      }),
    ]),
    'version': '0.0.0-test',
  })
# ---
# name: test_query_codebase_no_results
  dict({
    'results': list([
    ]),
    'version': '0.0.0-test',
  })
# ---
# name: test_status_endpoint_with_all_files_analyzed
  dict({
    'stats': dict({
      'accuracy': dict({
        'percentage': 100,
      }),
      'chunks': dict({
        'analyzed': 7,
        'unanalyzed': 0,
      }),
      'queue': dict({
        'size': 0,
      }),
    }),
    'version': '0.0.0-test',
  })
# ---
